src/java.base/share/classes/java/lang/reflect/Proxy.java

```diff
@@ -495,42 +495,23 @@ private static final class ProxyBuilder {
         private static final ClassLoaderValue<Boolean> reverseProxyCache =
             new ClassLoaderValue<>();
 
-        private static Class<?> defineProxyClass(Module m, List<Class<?>> interfaces) {
-            String proxyPkg = null;     // package to define proxy class in
-            int accessFlags = Modifier.PUBLIC | Modifier.FINAL;
-            boolean nonExported = false;
+        private record ProxyClassContext(Module module, String pkg, int accessFlags) { }
 
-            /*
-             * Record the package of a non-public proxy interface so that the
-             * proxy class will be defined in the same package.  Verify that
-             * all non-public proxy interfaces are in the same package.
-             */
-            for (Class<?> intf : interfaces) {
-                int flags = intf.getModifiers();
-                if (!Modifier.isPublic(flags)) {
-                    accessFlags = Modifier.FINAL;  // non-public, final
-                    String pkg = intf.getPackageName();
-                    if (proxyPkg == null) {
-                        proxyPkg = pkg;
-                    } else if (!pkg.equals(proxyPkg)) {
-                        throw new IllegalArgumentException(
-                                "non-public interfaces from different packages");
-                    }
-                } else {
-                    if (!intf.getModule().isExported(intf.getPackageName())) {
-                        // module-private types
-                        nonExported = true;
-                    }
-                }
-            }
+        private static Class<?> defineProxyClass(ProxyClassContext context, List<Class<?>> interfaces) {
+            String proxyPkg = context.pkg;
+            int accessFlags = context.accessFlags;
+            Module m = context.module;
 
-            if (proxyPkg == null) {
-                // all proxy interfaces are public and exported
+            if (Modifier.isPublic(accessFlags)) {
+                // all proxy interfaces are public
                 if (!m.isNamed())
                     throw new InternalError("unnamed module: " + m);
-                proxyPkg = nonExported ? PROXY_PACKAGE_PREFIX + "." + m.getName()
-                                       : m.getName();
-            } else if (proxyPkg.isEmpty() && m.isNamed()) {
+            }
+
+            if (proxyPkg.isEmpty() && m.isNamed()) {
+                // Per JLS 7.4.2, unnamed package can only exist in unnamed modules.
+                // This means a package-private superinterface exist in the unnamed
+                // package of a named module.
```

Comment:
With this patch, I think line 505-517 can turn into assert or internal error.   Probably can be moved to the `ProxyClassContext` constructor too.